Task 2

Saxophilyman opened the pull request.

Saxophilyman committed:
Create Java Project(empty)

Saxophilyman committed:
Part_I "Hello message"

Saxophilyman committed:
Part_II "First Steps(sketch)"

Saxophilyman committed:
Part_II "First Steps(sketch)"
Hope that this shit also added

Saxophilyman committed:
What is logic? (also add exit and repeat Begin in console)

Saxophilyman committed:
What is logic? (also add exit and repeat Begin in console)

Saxophilyman committed:
Part_III "Fucking Fuck"
Different lost ways

Saxophilyman committed:
Part_III "Details"
 - clear excess;
 - add readConsole;
 - add addCoord;
 - working on checkEndGame

Saxophilyman committed:
Part_III "Details 1.1" (thoughts about two player)

Saxophilyman committed:
Part_IV "Trying OOP"
 - rename class Begining to Game;
 - delete other classes;
 - clear Main(go in Game);
 - create class Player and his methods(+ 2 object);
 - some logic of game with class Player;

 - working on checkEndGame

Saxophilyman committed:
Part_V "EndProblem"
 - addedCord
 - notEqualsaddedCord

Saxophilyman committed:
Part_VI "Trying to be ready"
 - delete Player.java
 - true for checkEndGame
 - while & logic for Playing
 - Begin() in Main
 - circle gameGoOn(with Begin)
 - refreshGamePole

 - delete excess

Saxophilyman committed:
"Trying git_ignore"

Saxophilyman committed:
"Trying git_ignore 2.0"

Saxophilyman committed:
Part_VII "Trying to be ready 2.0"
 - check pull requests's comment #1

Saxophilyman committed:
Part_VII "Trying to be ready 2.1"
 - check pull requests's comment about decoration

Saxophilyman committed:
Part_VII "Trying to be ready 2.1.1"
 - check add Coord
 - problem with check EndGame in short way

Saxophilyman committed:
Part_VII "Trying to be ready 3.0"
- that's all

Saxophilyman committed:
Part_VIII "Trying to be ready 3.1"
 - with class Player

Saxophilyman committed:
Part_IX "3.2"
 - add constructor Game(Player, Player)
 - check mark

Saxophilyman committed:
Part_IX "3.3"
 - correct array (use new String[][])
 - correct printGameField(check NULL)
 - correct checkEndGame(reverse equals)

Saxophilyman committed:
пытаюсь добавить весь новый напечатанный код на гитхаб
- печатал код на компе
- перешёл на ноутбук
- установил IDE
- скопировал код
- редактировал на ноуте
- гит на ноуте не был добавлен
- пытаюсь добавить весь новый напечатанный код на гитхаб

Saxophilyman committed:
пытаюсь добавить весь новый напечатанный код на гитхаб
- печатал код на компе
- перешёл на ноутбук
- установил IDE
- скопировал код
- редактировал на ноуте
- гит на ноуте не был добавлен
- пытаюсь добавить весь новый напечатанный код на гитхаб

Saxophilyman committed:
Merge remote-tracking branch 'origin/task_2' into task_2

Saxophilyman committed:
Part_X "1.0"
- code style to Standart
- del some comment
- trying fix methods by settings Size of workField(Gor,Ver)

Saxophilyman committed:
Delete out/production/Krestiki_vs_Noliki-task_2 directory

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+
+    //ШАМАНСТВО С УСТАНОВКОЙ РАЗМЕРОВ ПОЛЯ
+    //Создание игрового поля
+    private static void createPlayFieldSize() {
```
почему метод статический? В этом нет необходимости.

Saxophilyman replied:
done

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+
+public class Game {
+    private static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+    static int Gor;
```
Прочитай про принцип инкапсуляции. Если вкратце - максимальное сокрытие реализации, в данном случае - переменные должны быть приватными, они ведьне используются за пределами класса. Плюс static здесь так же без необходимости.
Ну и название Gor, Ver - это названия гор в Тибете или что?:) horizontalFieldSize, verticalFieldSize.

Saxophilyman replied:
done

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+    }
+
+    //Запуск метода проверки диапазона поля от 3 до 10 и обработка неправильного значения
+    private static void CheckGorVer(int GorVer, String text, String s1, String s2) {
```
метод с маленькой буквы. Названия параметров так же переименуй на что-то более читаемое.:) Ну и само название метода лучше придумать попонятнее, например checkInputFieldSize. Передавать строки в этот метод, чтобы напечатать пару строчек - особо смысла нету, проще их после вызова метода напечатать отдельно.

Saxophilyman replied:
считаю что передачу строк в метод вообще можно убрать., т.к. по существу к логике задачи они не относятся. Но на данном этапе помогает легче ориентироваться и разграничить проверку на  ввод значений по горизонтали и вертикали. Предлагаю оставить в изначальном варианте.

epereskokov replied:
Ну вот именно, что к логике не относится и засоряет код. Логи подобного рода надо писать по месту, но передавать параметры для логов в методы - ну только если это действительно оочень нужно, лучше стараться этого не делать.

Saxophilyman replied:
Убрал всё лишнее по "логам" текста, оставил только самое необходимое
Done

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+    private static void CheckGorVer(int GorVer, String text, String s1, String s2) {
+        System.out.println("Получили " + GorVer);
+        while (!isValid(GorVer)) {
+            System.out.println("Отлично Блятъ! ");
```
перенос строк можно делать с помощью \n в тексте.

Saxophilyman replied:
done

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+    }
+
+    //Чтение параметров и запуск метода по считыванию размеров поля
+    private static int getFieldSize(String s1, String s2) {
```
передаваемым в метод параметрам надо давать читаемые имена, иначе непонятно, что еще за s1 и s2 ожидаются.

Saxophilyman replied:
done

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+
+    //Сама функция проверки диапазона считанного числа с консоли от 3 до 10
+    public static boolean isValid(int x) {
+        if (x > 2 && x < 11) {
```
public static boolean isValid(int size) {
return size >= 3 && size <=10;
}

Безымянные числа в коде ухудшают читаемость (называется magic numbers), поэтому подобные числовые константы обычно выносят в статичные переменные
private static final MIN_FIELD_SIZE = 3;
private static final MAX_FIELD_SIZE = 10;

Saxophilyman replied:
done
А почему тогда static?
и почему предпочтительнее "больше или равно" чем просто "больше/меньше"

Saxophilyman replied:
А почему тогда static?
и почему предпочтительнее "больше или равно" чем просто "больше/меньше"

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+    }
+
+    //Читаем координаты с консоли для постановки крестика-нолика
+    public void readConsole(Player player) {
```
как в комменте пишешь - так и назови.:) readCoordinateFromConsole

Saxophilyman replied:
done

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+            System.out.println("Извините, таких координат не существует, попробуйте ещё раз");
+            readConsole(player);
+        } else {
+            String gorCoord = String.valueOf(firstAndSecondCoordReadCrossZeroCord[0]);
```
пиши названия без сокращений horizontalCoordinate и тд

Saxophilyman replied:
done
time 1.5 hour

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+    //Проверка, были ли уже заданные данные координаты
+    private boolean checkIfNotEqualsAddedCord(String x) {
+        boolean check = true;
+        for (int i = 0; i < addedCoord.size(); i++) {
```
есть метод который делает как раз то, что тебе нужно addedCoord.contains(x)

опять непонятное названия параметра в методе

Saxophilyman replied:
done

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+        for (int idemPoStrokam = 0; idemPoStrokam < workField.length; idemPoStrokam++)
+            for (int idemPoStolbcam = 0; idemPoStolbcam < workField[idemPoStrokam].length; idemPoStolbcam++) {
+                if (idemPoStolbcam < workField[idemPoStrokam].length - 2) {
+                    if ((player.getMark().equals(workField[idemPoStrokam][idemPoStolbcam])) &&
```
у тебя кучу раз повторяется player.getMark(), сохрани в начале метода в переменную mark и везде ее используй.

Saxophilyman replied:
done

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+                            (player.getMark().equals(workField[idemPoStrokam][idemPoStolbcam + 1])) &&
+                            (player.getMark().equals(workField[idemPoStrokam][idemPoStolbcam + 2]))) {
+                        System.out.println("Поздравляем! " + player.getNameMark() + " Win!");
+                        gameNotEnd = false;
```
ты можешь сразу выйти из цикла с помощью break, раз ты уже выяснил, что условие выполнилось, тебе уже не нужно дальше все перебирать. Это же касается таких же мест ниже. А вообще можешь просто сразу вернуть из метода значение 
return false

Saxophilyman replied:
"А вообще можешь просто сразу вернуть из метода значение return false"
Как это работает? return сразу завершает цикл  и не доходит до последнего return?
есть идея. 
а по этому пункту -
done

epereskokov replied:
return завершает текущий метод и возвращает значение, которое ты передал. Соответственно, дальше метод не выполняется, так как ты уже вернул из него значение.

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+        boolean gameNotEnd = true;
+
+
+        for (int idemPoStrokam = 0; idemPoStrokam < workField.length; idemPoStrokam++)
```
idemPoStrokam - rowIndex
idemPoStolbcam - columnIndex

Saxophilyman replied:
мне было бы непонятно. предлагаю goByIndexOfLines/goByIndexOfColumns
если принимается - 
done

epereskokov replied:
переменная должны быть существительным. goByIndexOfLines - имеет глагол в названии. indexOfLines - ok

Saxophilyman replied:
done

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,358 @@
+                        gameNotEnd = false;
+                    }
+                }
+                if (idemPoStolbcam < workField[idemPoStrokam].length - 2 && idemPoStrokam < workField.length - 2) {
```
оригинальный алгоритм. Только читается плоховато из-за очень больших условий в if'ах.

Я предлагаю тебе написать рядом второй метод с другой логикой (а точнее - сразу 3 метода).
Первый будет по очереди проверять горизонтали, второй - вертикали, а третий - диагонали.
проверять будут, просто идя по очереди по элементам и считая непрерывное количество совпадающих с нужным значком элементов. Например, идем по первой строке, нашли крестик - увеличили счетчик совпадений до 1, проверили следующую ячейку - тоже крестик? Увеличил счетчик до 2, нолик? Сбросили до нуля и так далее.

Saxophilyman replied:
done

Saxophilyman committed:
Part_XI ""Trying to be ready 4.0""
 - check pull requests's comment (#2)

Saxophilyman committed:
Part_XI "Trying to be ready 4.1"
 - optimization of readConcoleFieldSize;
 - delete checkInputReadConsoleSizeForField;
 - correct getNameMark.

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,360 @@
+    //Чтение параметров и запуск метода по считыванию размеров поля
+    private static int getFieldSize(String indicatingOfLinesOrColumns, String indicatingOfHorizontalOrVertical) {
+        //BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+        System.out.println("Количество " + indicatingOfLinesOrColumns + " (" + indicatingOfHorizontalOrVertical + ")" + ":");
```
чтобы везде не печатать System.out.println(..), напиши отдельный маленький метод 
private void printMessage(String message) {
System.out.println(message);
}
Это сократит повторения в коде и улучшит читаемость.

Saxophilyman replied:
А какая существенная разница в этом? останутся повторения в стиле: 
printMessage("бла-бла");
верно? 
количество символов разнится на пару букв;
смысловое "sout" абсолютно ясное - не запутаешься;
также необходимо ставить кавычки;

сделать это не сложно, но совсем не понимаю в этом "облегчения".
пока оставлю на аргументацию)

epereskokov replied:
System.out.println - это читается хуже, чем printMessage. 
Ты сейчас потихоньку учишься писать в концепции "самодокументриующийся код". КОнцепция очень проста - твой код должен быть аки книга, которая не нуждается в комментариях. Мне проще прочитать printMessage, чем каждый раз читать длинное System.out.. System.out - какая-то там система, которая не имеет прямого отношения к логике программы, а значит при многократном повторении этот вызов можно спрятать в короткий метод с емким названием.

Понятно, что тут очень пограничный пример, в целом конечно восприятие не сильно страдает, если оставишь System.out, но тут важно именно то, чтобы ты понял саму идею самодокументриющегося кода, потому вызовы уже будут сложнее.:)

Saxophilyman committed:
Part_XI "Trying to be ready 4.3"
 - clear needless "textLog";
 - add method printMessage;
 - was corrected countSamePlayerMark;

Saxophilyman committed:
Forgot about this:
 Part_XI "Trying to be ready 4.2"
 - another method for "checkGameNotEnd" & methods for him;
 - checkEndGameByLines;
 - checkEndGameByColumns;
 - checkEndGameDiagonalFromLeftUpToRightDown;
 (cheking diagonal "twoDivMassive" from each index of Lines & Columns)
 - checkEndGameDiagonalFromLeftDownToRightUp(also).
 --------
 - checkThreeSameMarkInOneLine.
